src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/hmi/subscribe_button_request.cc

```diff
@@ -0,0 +1,163 @@
+                   application_manager,
+                   rpc_service,
+                   hmi_capabilities,
+                   policy_handle)
+    , EventObserver(application_manager.event_dispatcher()) {
+  hmi_apis::Common_ButtonName::eType button_name = static_cast<
+      hmi_apis::Common_ButtonName::eType>(
+      (*message_)[app_mngr::strings::msg_params][app_mngr::strings::button_name]
+          .asInt());
+
+  button_name_ = button_name;
+}
+
+SubscribeButtonRequest::~SubscribeButtonRequest() {
+  unsubscribe_from_event(hmi_apis::FunctionID::Buttons_SubscribeButton);
+}
+
+void SubscribeButtonRequest::Run() {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  subscribe_on_event(hmi_apis::FunctionID::Buttons_SubscribeButton,
+                     correlation_id());
+
+  ApplicationSharedPtr app =
+      application_manager_.application_by_hmi_app(application_id());
+
+  if (!app) {
+    LOG4CXX_ERROR(logger_,
+                  "Application for connection key: " << application_id()
+                                                     << " was not found");
+    return;
+  }
+
+  app->AddPendingSubscriptionButton(correlation_id(), button_name_);
+
+  SendRequest();
+}
+
+void SubscribeButtonRequest::onTimeOut() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  application_manager_.updateRequestTimeout(
+      connection_key(), correlation_id(), 0);
+
+  auto& resume_ctrl = application_manager_.resume_controller();
+
+  ApplicationSharedPtr app =
+      application_manager_.application_by_hmi_app(application_id());
+
+  if (!app) {
+    LOG4CXX_ERROR(logger_,
+                  "Application for app id: " << application_id()
+                                             << " was not found");
+    return;
+  }
+
+  app->RemovePendingSubscriptionButton(correlation_id());
+
+  resume_ctrl.HandleOnTimeOut(
+      correlation_id(),
+      static_cast<hmi_apis::FunctionID::eType>(function_id()));
+}
+
+void SubscribeButtonRequest::on_event(const event_engine::Event& event) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  using namespace helpers;
+
+  const smart_objects::SmartObject& message = event.smart_object();
+
+  if (hmi_apis::FunctionID::Buttons_SubscribeButton != event.id()) {
+    LOG4CXX_ERROR(logger_, "Received unknown event.");
+    return;
+  }
+
+  const uint32_t app_id =
+      (*message_)[strings::msg_params][strings::app_id].asUInt();
+
+  ApplicationSharedPtr app =
+      application_manager_.application_by_hmi_app(app_id);
+
+  if (!app) {
+    LOG4CXX_ERROR(logger_, "NULL pointer.");
+    return;
+  }
+
+  hmi_apis::Common_Result::eType hmi_result =
+      static_cast<hmi_apis::Common_Result::eType>(
+          message[strings::params][hmi_response::code].asInt());
+
+  const bool is_in_pending = app->PendingSubscriptionButtons().find(
+                                 event.smart_object_correlation_id()) !=
+                             app->PendingSubscriptionButtons().end();
+
+  if (hmi_apis::Common_Result::SUCCESS == hmi_result && is_in_pending) {
```

Comment:
@mked-luxoft What if HMI result will be some other positive like `WARNING` ?

Reply:
@AByzhynar fixed in2b001c6014447a350d41a2bcd1e834fd2c83057a